Junior services (#539)
gavofyork/graypaper: text/accumulation.tex

\newcommand*{\local¬fngasused}{U}
\newcommand*{\local¬fnprovidable}{Y}
\newcommand*{\local¬fnprovide}{I}

> \newcommand*{\local¬fnpartitionxfers}{T}

Reviewer: Nit. $T()$ collides with the statistics function in the same section.
Same for $N()$, but I think this is fine as it is defined within the context of $\Delta_*$

Author: Yeah ok. I'll rename one of them...

Author: Renamed the statistics one to X.

As the result of accumulating a transfer cannot reasonably be discarded (this would cause funds to be lost), deferred transfers to a service are \emph{never} accumulated in parallel with execution of the Accumulate logic of the supervisor. We define the function $T$ which partitions a sequence of deferred transfers into those which may be accumulated in parallel with the Accumulate logic of the services $\mathbf{s}$ and those which must be accumulated in a later round.
\begin{equation}
  \local¬fnpartitionxfers\colon\abracegroup{
    &\tuple{\sequence{\defxfer}, \protoset{\serviceid}, \dictionary{\serviceid}{\serviceaccount}} \to \tuple{\sequence{\defxfer}, \sequence{\defxfer}} \\
    &\tup{\sq{}, \mathbf{s}, \mathbf{d}} \mapsto \tup{\sq{}, \sq{}} \\
    &\tup{\sq{t} \concat \mathbf{t}, \mathbf{s}, \mathbf{d}} \mapsto \begin{cases}
      \tup{\mathbf{t}^!, \sq{t} \concat \mathbf{t}^Q} &\when d = \top \\
      \tup{\sq{t} \concat \mathbf{t}^!, \mathbf{t}^Q} &\otherwise
    \end{cases} \\

>     &\quad\where d = t_\dx¬dest \in \mathbf{d} \land \effectivesupervisor{\mathbf{d}}{t_\dx¬dest} \in \mathbf{s} \setminus \set{t_\dx¬dest} \\

Author: This check is not sufficient... `t_d` might be the supervisor of some service in `\mathbf{s}`, this only checks if `t_d`'s supervisor is in `\mathbf{s}`.

Author: Fixed...

  \local¬fngasused \equiv \sequence{\tuple{\serviceid, \gas}}
\end{equation}

> It is possible for a service's state to be modified by both the service's Accumulate logic and its supervisor's Accumulate logic. We wish to avoid inconsistent state resulting from concurrent modification while not unnecessarily restricting parallel accumulation. To this end, we optimistically permit execution of a service's Accumulate logic and its supervisor's Accumulate logic in the same round, but discard all results\footnote{With one exception: provided preimages are not discarded} of the service's execution if both modify the service's state. Note that ejection of a service is considered a modification of its state.

Reviewer: ```suggestion
It is possible for a service's state to be modified by both the service's Accumulate logic and its supervisor's Accumulate logic. We wish to avoid inconsistent state resulting from concurrent modification while not unnecessarily restricting parallel accumulation. To this end, we optimistically permit execution of a service's Accumulate logic and its supervisor's Accumulate logic in the same round, but discard all results\footnote{With one exception: provided preimages are not discarded} of the service's execution if both modify the service's state. Note that ejection of a service is considered a modification of its state. A service scheduled for accumulation may itself have been ejected in an earlier round, and thus no longer exist. Such a service accumulates to nothing and consumes no gas.
```

Nit for the new guard https://github.com/gavofyork/graypaper/pull/539#discussion_r3740341753

Author: I've not added this as I've not added the guard (I think it's redundant). Maybe it's worth calling out here what happens when a non-existent service is accumulated though :thinking:

> As the result of accumulating a transfer cannot reasonably be discarded (this would cause funds to be lost), deferred transfers to a service are \emph{never} accumulated in parallel with execution of the Accumulate logic of the supervisor. We define the function $T$ which partitions a sequence of deferred transfers into those which may be accumulated in parallel with the Accumulate logic of the services $\mathbf{s}$ and those which must be accumulated in a later round.

Reviewer: ```suggestion
As the result of accumulating a transfer cannot reasonably be discarded (this would cause funds to be lost), deferred transfers to a service are \emph{never} accumulated in parallel with execution of the Accumulate logic of the supervisor. We define the function $\local¬fnpartitionxfers$ which partitions a sequence of deferred transfers into those which may be accumulated in parallel with the Accumulate logic of the services $\mathbf{s}$ and those which must be accumulated in a later round.
```

\begin{equation}
  \begin{gathered}
    \local¬fnpartitionxfers(\mathbf{t}, \mathbf{s}, \mathbf{d}) \equiv \local¬fnpartitionxfers^*(\mathbf{t}, \mathbf{s}, \emptyset, \mathbf{d}) \\
    \local¬fnpartitionxfers^*\colon\abracegroup{
      &\tuple{\sequence{\defxfer}, \protoset{\serviceid}, \protoset{\serviceid}, \dictionary{\serviceid}{\serviceaccount}} \to \tuple{\sequence{\defxfer}, \sequence{\defxfer}} \\
      &\tup{\sq{}, \mathbf{s}, \mathbf{i}, \mathbf{d}} \mapsto \tup{\sq{}, \sq{}} \\
      &\tup{\sq{t} \concat \mathbf{t}, \mathbf{s}, \mathbf{i}, \mathbf{d}} \mapsto \begin{cases}
        \tup{\mathbf{t}^!, \sq{t} \concat \mathbf{t}^Q} &\when q = \top \\
        \tup{\sq{t} \concat \mathbf{t}^!, \mathbf{t}^Q} &\otherwise
      \end{cases} \\

>       &\quad\where q = (t_\dx¬dest \in \mathbf{d} \land \effectivesupervisor{\mathbf{d}}{t_\dx¬dest} \in (\mathbf{s} \cup \mathbf{i}) \setminus \set{t_\dx¬dest}) \lor {} \\

Reviewer: ```suggestion
      &\quad\where q = (t_\dx¬dest \in \keys{\mathbf{d}} \land \effectivesupervisor{\mathbf{d}}{t_\dx¬dest} \in (\mathbf{s} \cup \mathbf{i}) \setminus \set{t_\dx¬dest}) \lor {} \\
```

      &\quad\phantom{\where q = {}} \exists s \in \mathbf{i} \setminus \set{t_\dx¬dest} : \effectivesupervisor{\mathbf{d}}{s} = t_\dx¬dest \\
      &\quad\also \mathbf{i}' = \begin{cases}

>         \mathbf{i} &\when q = \top \lor t_\dx¬dest \not\in \mathbf{d} \\

Reviewer: ```suggestion
        \mathbf{i} &\when q = \top \lor t_\dx¬dest \not\in \keys{\mathbf{d}} \\
```

    }\!\!\!\!\!\!\\
    &\text{where:}\\
    &\ \begin{aligned}
      \using &\tup{
        \ps¬accounts, \ps¬stagingset, \ps¬authqueue, \ps¬manager, \ps¬assigners, \ps¬delegator, \ps¬registrar, \ps¬alwaysaccers
      } = \psX \\
      \mathbf{s} &= \set{\build{
        d_\wd¬serviceindex
        }{
          r \in \mathbf{r}, d \in r_\wr¬digests
        }} \cup \keys{\mathbf{f}} \cup \set{\build{t_\dx¬dest}{t \in \mathbf{t}}} \\
      N(\mathbf{o}) \in \acconeout &\equiv \tup{
        \ao¬poststate,
        \is{\ao¬modifiedaccounts}{\emset},
        \is{\ao¬defxfers}{\sq{}},
        \is{\ao¬yield}{\none},
        \is{\ao¬gasused}{\mathbf{o}_\ao¬gasused},
        \is{\ao¬provisions}{\mathbf{o}_\ao¬provisions}
      } \\
      \local¬acc(s) &\equiv \begin{cases}

>         N(\mathbf{o}) &\when s \in \mathbf{o}_\ao¬modifiedaccounts \land s \in \mathbf{y}_\ao¬modifiedaccounts \setminus \set{\effectivesupervisor{\ps¬accounts}{s}} \\

Reviewer: ```suggestion
        \mathbf{o} &\when s \not\in \keys{\ps¬accounts} \\
        N(\mathbf{o}) &\otherwhen s \in \mathbf{o}_\ao¬modifiedaccounts \land s \in \mathbf{y}_\ao¬modifiedaccounts \setminus \set{\effectivesupervisor{\ps¬accounts}{s}} \\
```

Author: I didn't bother with an explicit check because `s \in \mathbf{o}_\ao¬modifiedaccounts` is well-defined but always false when the service `s` does not exist.

        \mathbf{o} &\otherwise
      \end{cases} \\
      &\phantom{{}\equiv{}} \where \mathbf{o} = \accone(\psX, \mathbf{t}, \mathbf{r}, \mathbf{f}, s) \\

>       &\phantom{{}\equiv{}} \also \mathbf{y} = \accone(\psX, \mathbf{t}, \mathbf{r}, \mathbf{f}, \effectivesupervisor{\ps¬accounts}{s}) \\

Reviewer: The "nullification" test uses the supervisor's raw execution, not its post-discard one. E.g. with e(a)=b, e(b)=c: a writes a, b writes a and b, c writes b , then b is nullified by c, but a is still nullified by b's write to a, which never lands.  I imagine that this is deliberate and is not worth over-complicating

Author: Yeah it is intentionally this way for simplicity. You could define something more complicated which would eg in your case only nullify b. But cases like this ought to be rare, so it doesn't seem worth the effort. Note that you can't just say "use the post-discard result" because that would be undefined in the case of authority loops (which while obviously not expected are strictly speaking possible to setup and so need well-defined behaviour).

\begin{gather}
  R_D(\mathbf{a}, \mathbf{w}, \mathbf{b}) \equiv \set{\build{\kv{\mathbf{k}}{\mathbf{v}}}{\kv{\mathbf{k}}{\mathbf{v}} \in \mathbf{a}, \mathbf{k} \in \mathbf{w}}} \cup (\mathbf{b} \setminus \mathbf{w}) \\
  R_P(\mathbf{p}, \mathbf{l}) \equiv \set{\build{\kv{h}{\mathbf{p}\subb{h}}}{\kv{\tup{h, z}}{\mathbf{r}} \in \mathbf{l}, \mathbf{r} \ne \sq{}}} \\
  R_A^\dagger(\mathbf{o}, \mathbf{a}, \mathbf{w}, \mathbf{r}, \mathbf{b}) \equiv \tup{\begin{aligned}
    &\is{\sa¬codehash}{R(\mathbf{o}_\sa¬codehash, \mathbf{a}_\sa¬codehash, \mathbf{b}_\sa¬codehash)},
      \is{\sa¬storage}{R_D(\mathbf{a}_\sa¬storage, \mathbf{w}, \mathbf{b}_\sa¬storage)}, \\
    &\is{\sa¬requests}{R_D(\mathbf{a}_\sa¬requests, \mathbf{r}, \mathbf{b}_\sa¬requests)},
      \is{\sa¬balance}{\mathbf{b}_\sa¬balance},
      \is{\sa¬authority}{\mathbf{a}_\sa¬authority},

>       \is{\sa¬authbalance}{\mathbf{a}_\sa¬authbalance}, \\

Reviewer: Taking $\mathbf{a}_x$ alone drops deferred transfers with destauth=⊤ : $Ψ_A$ credits them into the destination's accumulation context, so they only appear in $\mathbf{b}_x$ and thus the authority's context never sees them (it only receives transfers destined to itself). So when a service's authority is not itself, the credit is discarded at the merge while the sender's debit (made at transfer call time) persists.

Now we have $x = \mathbf{a}_x$ which is  only the authority's view; deferred credits (which live in $\mathbf{b}_x$) are lost.
A fix can be $x = a_x + b_x - o_x$ : the authority's view plus the delta from the service's own run, which can only be deferred credits. If nothing the $\mathbf{b}_x = \mathbf{o}_x$

Author: Hmm yeah thanks, good spot. Note that `a` and `b` can be from the same service so `a_x + b_x - o_x` doesn't work... in the case they are the same you would double any change! I can of course detect that case and make it work. Perhaps transfers to the authority balance should be processed by the authority service anyway though? Any opinion here @gavofyork?

Reviewer: Just for clarity... Funds are lost exactly when (in `transfer` hostcall) $\mathbf{o} \neq 0$ and $x \in {2,3}$ and dest's authority not equal dest: the `transfer` hostcall enqueues a deferred transfer for `dest`, which is delivered to dest's own accumulation (so credit lands in **b**), but R_A† takes authbalance from **a**, which never sees the transfer

Author: Should be fixed now. I've stuck with transfers to the authority balance being handled by the service rather than its authority service. The authority balance deltas are simply added as you proposed, with a check in $R_A$ to handle the case where the service and its authority service are actually the same.

Reviewer 2: > Hmm yeah thanks, good spot. Note that `a` and `b` can be from the same service so `a_x + b_x - o_x` doesn't work... in the case they are the same you would double any change! I can of course detect that case and make it work. Perhaps transfers to the authority balance should be processed by the authority service anyway though? Any opinion here @gavofyork?

As discussed I think it makes more sense to keep all transfers be processed by the destination service.
